language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/ExBudgeting.hs

```diff
@@ -178,34 +181,45 @@ estimateStaticStagedCost
-    , paramSHA3                 :: CostingFun ModelOneArgument
-    , paramVerifySignature      :: CostingFun ModelThreeArguments
-    , paramEqByteString         :: CostingFun ModelTwoArguments
-    , paramLtByteString         :: CostingFun ModelTwoArguments
-    , paramGtByteString         :: CostingFun ModelTwoArguments
-    , paramIfThenElse           :: CostingFun ModelThreeArguments
+    { paramAddInteger           :: f ModelTwoArguments
+    , paramSubtractInteger      :: f ModelTwoArguments
+    , paramMultiplyInteger      :: f ModelTwoArguments
+    , paramDivideInteger        :: f ModelTwoArguments
+    , paramQuotientInteger      :: f ModelTwoArguments
+    , paramRemainderInteger     :: f ModelTwoArguments
+    , paramModInteger           :: f ModelTwoArguments
+    , paramLessThanInteger      :: f ModelTwoArguments
+    , paramLessThanEqInteger    :: f ModelTwoArguments
+    , paramGreaterThanInteger   :: f ModelTwoArguments
+    , paramGreaterThanEqInteger :: f ModelTwoArguments
+    , paramEqInteger            :: f ModelTwoArguments
+    , paramConcatenate          :: f ModelTwoArguments
+    , paramTakeByteString       :: f ModelTwoArguments -- TODO these two might be a bit interesting on size
+    , paramDropByteString       :: f ModelTwoArguments
+    , paramSHA2                 :: f ModelOneArgument
+    , paramSHA3                 :: f ModelOneArgument
+    , paramVerifySignature      :: f ModelThreeArguments
+    , paramEqByteString         :: f ModelTwoArguments
+    , paramLtByteString         :: f ModelTwoArguments
+    , paramGtByteString         :: f ModelTwoArguments
+    , paramIfThenElse           :: f ModelThreeArguments
     }
-    deriving (Show, Eq, Generic, Lift, Default, NFData)
-    deriving (FromJSON, ToJSON) via CustomJSON '[FieldLabelModifier (StripPrefix "param", CamelToSnake)] CostModel
+    deriving (Generic, FunctorB, TraversableB, ConstraintsB)
+
+deriving via CustomJSON '[FieldLabelModifier (StripPrefix "param", CamelToSnake)] (CostModelBase CostingFun) instance ToJSON (CostModelBase CostingFun)
+deriving via CustomJSON '[FieldLabelModifier (StripPrefix "param", CamelToSnake)] (CostModelBase CostingFun) instance FromJSON (CostModelBase CostingFun)
+
+-- HLS doesn't like the AllBF from Barbies.
+deriving instance (NFData (f ModelTwoArguments), NFData (f ModelThreeArguments), NFData (f ModelOneArgument)) => NFData (CostModelBase f)
+deriving instance (Default (f ModelTwoArguments), Default (f ModelThreeArguments), Default (f ModelOneArgument)) => Default (CostModelBase f)
+deriving instance (Lift (f ModelTwoArguments), Lift (f ModelThreeArguments), Lift (f ModelOneArgument)) => Lift (CostModelBase f)
+deriving instance (Show (f ModelTwoArguments), Show (f ModelThreeArguments), Show (f ModelOneArgument)) => Show (CostModelBase f)
+deriving instance (Eq (f ModelTwoArguments), Eq (f ModelThreeArguments), Eq (f ModelOneArgument)) => Eq   (CostModelBase f)
```

Comment:
You could use `GEq`, `GShow` and other stuff from the `dependent-sum` package. And maybe even some of our stuff from `Language.PlutusCore.Universe.Core`, but I think it would be an overkill and I like your version more. I'd try to abstract the `(constr (f ModelTwoArguments), constr (f ModelThreeArguments), constr (f ModelOneArgument))` pattern, though.

Reply:
Forgot about constraint kinds, thanks!